tempesta_fw/apm.c

```diff
@@ -0,0 +1,1039 @@
+ */
+static int
+tfw_apm_calc(TfwApmData *data)
+{
+	int ret, wridx;
+	Percentile percentile[] = {
+		{tfw_apm_prcntl_ith[0]}, {tfw_apm_prcntl_ith[1]},
+		{tfw_apm_prcntl_ith[2]}, {tfw_apm_prcntl_ith[3]},
+		{tfw_apm_prcntl_ith[4]}, {tfw_apm_prcntl_ith[5]},
+	};
+	TfwApmPrcntl wrkprcntl = { percentile, ARRAY_SIZE(percentile) };
+	TfwApmPrcntl *prcntl;
+
+	BUILD_BUG_ON(ARRAY_SIZE(percentile) != ARRAY_SIZE(tfw_apm_prcntl_ith));
+
+	ret = __tfw_apm_calc(data, &wrkprcntl);
+
+	wridx = ((unsigned int)atomic_read(&data->stats.rdidx) + 1) % 2;
+	prcntl = &data->stats.prcntl[wridx];
+
+	write_lock(&prcntl->rwlock);
+	if (ret) {
+		TFW_DBG3("%s: Percentile values DID NOT change.\n", __func__);
+		prcntl->flags |= TFW_APM_DATA_F_SAMEOLD;
+	} else {
+		TFW_DBG3("%s: Percentile values may have changed.\n", __func__);
+		prcntl->flags &= ~TFW_APM_DATA_F_SAMEOLD;
+		memcpy(prcntl->percentile, percentile,
+		       prcntl->prcntlsz * sizeof(Percentile));
+		atomic_inc(&data->stats.rdidx);
+	}
+	write_unlock(&prcntl->rwlock);
+
+	return ret;
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ * Runs periodically on timer. 
+ */
+static void
+tfw_apm_prcntl_fn(unsigned long fndata)
+{
+	TfwApmData *data = (TfwApmData *)fndata;
+
+	tfw_apm_calc(data);
+
+	smp_mb();
+
+	if (test_bit(TFW_APM_DATA_F_REARM, &data->flags))
+		mod_timer(&data->timer, jiffies + TFW_APM_TIMER_TIMEOUT);
+}
+
+/*
+ * Get the latest calculated percentiles.
+ *
+ * Return 0 if potentially new percentile values were calculated.
+ * Return 1 if the percentile values didn't need recalculation.
+ */
+int
+tfw_apm_stats(void *apmdata, Percentile *percentile, size_t prcntlsz)
+{
+	int ret, rdidx;
+	TfwApmData *data = apmdata;
+	TfwApmPrcntl *prcntl;
+
+	BUG_ON(!apmdata);
+
+	rdidx = (unsigned int)atomic_read(&data->stats.rdidx) % 2;
+	prcntl = &data->stats.prcntl[rdidx];
+	if (!read_trylock(&prcntl->rwlock)) {
+		rdidx = (unsigned int)atomic_read(&data->stats.rdidx) % 2;
+		prcntl = &data->stats.prcntl[rdidx];
+		read_lock(&prcntl->rwlock);
+	}
+	memcpy(percentile, prcntl->percentile, prcntlsz * sizeof(Percentile));
+	ret = (prcntl->flags & TFW_APM_DATA_F_SAMEOLD);
+	read_unlock(&prcntl->rwlock);
+
+	return ret;
+}
+
+/*
+ * Verify that an APM Stats user using the same set of percentiles.
+ *
+ * Note: This module uses a single set of percentiles for all servers.
+ * All APM Stats users must use the same set of percentiles.
+ */
+int
+tfw_apm_percentile_verify(Percentile *prcntl, size_t prcntlsz)
+{
+	int i;
+
+	if (prcntlsz != ARRAY_SIZE(tfw_apm_prcntl_ith))
+		return 1;
+	for (i = 0; i < prcntlsz; ++i)
+		if (prcntl[i].ith != tfw_apm_prcntl_ith[i])
+			return 1;
+	return 0;
+}
+
+/*
+ * Initialize a ring buffer entry.
+ */
+static inline void
+tfw_apm_rbent_init(TfwApmRBufEnt *rbent, unsigned long jtstamp)
+{
+	memset(&rbent->pcntrng, 0, sizeof(TfwPcntRanges));
+	memcpy(rbent->pcntrng.ctl, tfw_rngctl_init, sizeof(rbent->pcntrng.ctl));
```

Comment:
I think we shouldn't initialize the ranges by `tfw_rngctl_init`. We've learnt the appropriate ranges and typically they aren't changed from time to time. Meantime we loose accuracy when we adjust ranges. So it's better to improve accuracy when the system learn the ranges.